### PR TITLE
[Dataset] Add ZINC Dataset

### PR DESCRIPTION
**Pull request by ZhenyuLU-Heliodore**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by rudongyu** [comment]
@dgl-bot

File: docs/source/api/python/dgl.data.rst [hunk 1]
```diff
@@ -55,6 +55,7 @@ Datasets for node classification/regression tasks
     FlickrDataset
     YelpDataset
     CLUSTERDataset
+    ZINCDataset
```
Comment (rudongyu) [inline docs/source/api/python/dgl.data.rst:58]: Note that ZINC is for graph prediction.

File: python/dgl/data/zinc.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
+import os
+
+from .dgl_dataset import DGLBuiltinDataset
+from .utils import _get_dgl_url, load_graphs
+
+
+class ZINCDataset(DGLBuiltinDataset):
+    r""" ZINC dataset for graph regression task.
```
Comment (rudongyu) [inline python/dgl/data/zinc.py:8]:
```suggestion
    r""" ZINC dataset for the graph regression task.
```

File: python/dgl/data/zinc.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+   num_atom_types : int
+       Number of atom types.
+   num_bond_types : int
+       Number of bond types.
+
+   Examples
+   ---------
+   >>> from dgl.data import ZINCDataset
+   >>>
+   >>> training_set = ZINCDataset(mode="train")
```
Comment (rudongyu) [inline python/dgl/data/zinc.py:58]:
```suggestion
   >>> from dgl.data import ZINCDataset

   >>> training_set = ZINCDataset(mode="train")
```

File: python/dgl/data/zinc.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+   >>> training_set.num_atom_types
+   28
+   >>> len(training_set)
+   10000
+   >>> graph, label = training_set[0]
+   >>> graph
+   Graph(num_nodes=29, num_edges=64,
+         ndata_schemes={'feat': Scheme(shape=(), dtype=torch.int64)}
+         edata_schemes={'feat': Scheme(shape=(), dtype=torch.int64)})
+   """
```
Comment (rudongyu) [inline python/dgl/data/zinc.py:68]: please check the indent of documentation.

File: tests/python/common/data/test_data.py [hunk 5]
```diff
@@ -386,6 +386,31 @@ def test_cluster():
...
+    }
+    transform = dgl.AddSelfLoop(allow_duplicate=True)
+    for mode, n_graphs in mode_n_graphs.items():
+        dataset1 = data.ZINCDataset(mode=mode)
+        g1, label = dataset1[0]
+        dataset2 = data.ZINCDataset(mode=mode, transform=transform)
+        g2, _ = dataset2[0]
+
+        assert g2.num_edges() - g1.num_edges() == g1.num_nodes()
+        assert dataset2.num_atom_types == 28
```
Comment (rudongyu) [inline tests/python/common/data/test_data.py:410]: The static property of the dataset don't need to be tested.

**Review by rudongyu — CHANGES_REQUESTED** [review]

File: python/dgl/data/zinc.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        idx : int
+            The sample index.
+
+        Returns
+        -------
+        dgl.DGLGraph
+            Each graph contains:
+
+            - ``ndata['feat']``: Types of heavy atoms as node features
+            - ``edata['feat']``: Type of bonds as edge features
```
Comment (rudongyu) [inline python/dgl/data/zinc.py:128]:
```suggestion
            - ``edata['feat']``: Types of bonds as edge features
```

File: tests/python/common/data/test_data.py [hunk 7]
```diff
@@ -408,6 +408,30 @@ def test_cluster():
...
+    }
+    transform = dgl.AddSelfLoop(allow_duplicate=True)
+    for mode, n_graphs in mode_n_graphs.items():
+        dataset1 = data.ZINCDataset(mode=mode)
+        g1, label = dataset1[0]
+        dataset2 = data.ZINCDataset(mode=mode, transform=transform)
+        g2, _ = dataset2[0]
+
+        assert g2.num_edges() - g1.num_edges() == g1.num_nodes()
+        assert label.shape == 1
```
Comment (rudongyu) [inline tests/python/common/data/test_data.py:432]: Please recheck the assertion on label shape (label.shape returns a torch.Size object).

**Review by rudongyu — COMMENTED** [review]
Good job, thanks. please fix issues corresponding to the new comments.

File: python/dgl/data/zinc.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+        Tensor
+            Constrained solubility as graph label
+
```
Comment (frozenbugs) [inline python/dgl/data/zinc.py:132]: nit: redundant blank line.

**Comment by frozenbugs** [comment]
LGTM, @rudongyu feel free to merge.

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Review by rudongyu — APPROVED** [review]

**Comment by rudongyu** [comment]
@dgl-bot

**Pull request merged**